### ISSUE TITLE
[AI-ENGINE.12b] Diagnóstico opt-in de activaciones/logits por layer sin tocar el hot path

### ISSUE THREAD

**Issue by kalekoi** [AuriaLABS/Auralis#222]
## Parent
Part of #90 / #27 / #4. Governed by #25/#85.

## Estado
Preparado pero **no reclamado** mientras #218/#219 ocupan el WIP activo.

## Motivo
#219 cubre loss + gradients + parameters + Adam state. Falta activaciones/logits por layer.

No queremos añadir ramas de diagnóstico a `Gpt::logits()`, `forward_eval()` ni al entrenamiento normal.

## Diseño recomendado
Añadir una API debug separada que use el ya existente `forward_internal(tokens)` únicamente cuando el usuario pide diagnóstico.

`forward_internal` ya materializa los caches de training:
- h1
- q/k/v
- attention probs/att
- h2
- ff_pre
- ff_act
- h_final
- logits

La API diagnóstica puede escanear esos tensores después de calcularlos, sin cambiar el orden matemático.

## API orientativa
- `Gpt::forward_diagnostics(tokens, Diagnostics) -> Result<ForwardDiagnosticsReport, String>`
- `Diagnostics::off()` puede rechazar/no producir report o delegar a logits según contrato final;
- `Diagnostics::on()` usa `forward_internal`, recorre caches en orden estable y reporta primera corrupción.

## Orden de stage recomendado
1. embedding/residual input cuando sea observable sin duplicar lógica;
2. layer N / ln1 output (h1);
3. layer N / q;
4. layer N / k;
5. layer N / v;
6. layer N / attention probabilities;
7. layer N / attention output;
8. layer N / ln2 output (h2);
9. layer N / ff_pre;
10. layer N / ff_act;
11. final layernorm output;
12. logits.

## Checklist
- [ ] definir ForwardTensorSummary con layer/stage/name/Scan
- [ ] ordenar determinísticamente los summaries
- [ ] incluir índice de layer en errores
- [ ] detectar NaN/Inf con índice dentro del tensor
- [ ] escanear logits
- [ ] escanear h1/h2
- [ ] escanear q/k/v
- [ ] escanear attention probs/att
- [ ] escanear ff_pre/ff_act
- [ ] escanear h_final
- [ ] test pipeline limpio
- [ ] fixture NaN/Inf con nombre de layer/tensor
- [ ] primera corrupción gana
- [ ] comparar logits debug vs `Gpt::logits()` bit-for-bit
- [ ] benchmark coste debug
- [ ] demostrar que `Gpt::logits()` normal no cambia
- [ ] no añadir campo a RunConfig en esta slice

## Invariantes
- no cambio matemático;
- no modificación del hot path normal si puede evitarse;
- no unsafe;
- no cambio de checkpoint;
- la ruta debug puede ser allocation-heavy: es observabilidad opt-in;
- usar #29/#27 como oráculo de equivalencia inference/training.

## Criterio de aceptación
Un usuario puede ejecutar un forward diagnóstico y obtener la primera activación/logit no finita con layer+tensornombre+índice, manteniendo `Gpt::logits()` y el entrenamiento normal sin semántica nueva ni overhead añadido.

Después de esta slice, #90 solo debería necesitar decidir la superficie final de activación (CLI/debug config) y cerrar la medición end-to-end completa.

**Comment by kalekoi** [comment]
IMPLEMENTATION NOTE — zero hot-path impact

Recommended concrete shape:

- keep `forward_eval()` and `Gpt::logits()` byte-for-byte untouched;
- add a separate debug-only public method on Gpt that calls existing `forward_internal(tokens)`;
- while the returned ForwardCache is alive, scan its Vec slices directly; do not clone tensors into the report;
- report stores only compact summaries (layer index, tensor name, Scan), not activation payloads;
- scan order must be deterministic and match forward order;
- logits returned by forward_internal must be asserted bit-for-bit equal to Gpt::logits() in tests;
- first fault is derived from the ordered compact summaries;
- debug allocations are acceptable, but normal eval/inference must show identical allocations to pre-#222.

Suggested summary structure:
`ForwardTensorSummary { layer: Option<usize>, name: String, scan: Scan }`
plus `ForwardDiagnosticsReport { tensors: Vec<ForwardTensorSummary> }`.

Avoid adding `Diagnostics` branches inside each layer operation. The opt-in boundary is the separate method itself.

**Comment by kalekoi** [comment]
SURFACE NOTE

Current main does not wire `numeric::cli` into the top-level Auralis CLI. #222 should not assume that debug surface already exists.

After the forward report API is stable, the remaining #90 surface can either:
- add a top-level `auralis numeric ...` debug command; or
- add an explicit diagnostics/debug subcommand/config object.

Do not silently add diagnostics fields to persisted RunConfig unless the project decides observability should become reproducible run semantics.

**Comment by kalekoi** [comment]
CONCRETE IMPLEMENTATION CONTRACT — based on current model.rs

Current `ForwardCache` already gives all tensors needed without changing math:
- per layer: h1, q, k, v, probs, att, h2, ff_pre, ff_act;
- final: h_final;
- `forward_internal` returns logits separately.

Recommended report ordering per layer:
`h1 → q → k → v → probs → att → h2 → ff_pre → ff_act`,
then `h_final → logits`.

Do not scan `LnCache::xhat` in the first slice unless needed for fault localization; h1/h2 are the actual affine layernorm outputs consumed by the following ops. Keeping the first report compact reduces debug overhead and surface area.

Suggested API:
`pub fn forward_diagnostics(&self, tokens: &[usize]) -> Result<(Vec<f32>, ForwardDiagnosticsReport), String>`

The method itself is the opt-in boundary; no Diagnostics branch is required inside layer ops. It can:
1. call `forward_internal(tokens)`;
2. scan cached slices directly into compact summaries;
3. scan returned logits;
4. return first-fault error or clean report + logits.

Tests:
- clean report tensor count = `n_layer * 9 + 2`;
- deterministic name/order;
- clean debug logits == `self.logits(tokens)` bit-for-bit;
- corrupt-model fixture should identify earliest affected layer/tensor, but avoid mutating private caches just for a synthetic test if parameter corruption can naturally propagate;
- normal `logits()` allocation/profile baseline must remain unchanged because its code path is untouched.

This keeps #222 to model.rs + a small diagnostics type/helper module/test/benchmark surface and avoids RunConfig/checkpoint changes.

**Comment by kalekoi** [comment]
CLAIM
Agent: ChatGPT
Role: Implementer + Verifier + Benchmark
Base: main @ 1271bc14
Scope: model diagnostics API + compact summaries/tests/benchmark; no changes to normal logits/forward_eval path, RunConfig, checkpoint or math
Expected handoff: per-layer activation/logit scan, first-fault context, bit-for-bit debug logits equivalence, debug overhead measurement, normal-path unchanged
Conflicts checked: #223 touches benchmark-only matmul files; no overlap expected.

**Comment by kalekoi** [comment]
DONE — PR #226 merged as `272c2e4f`.

Delivered:
- debug-only forward diagnostics over cached activations/logits;
- deterministic per-layer summary order;
- first-fault context with layer/tensor/index;
- exact debug logits == normal logits on clean runs;
- no normal inference/eval hot-path branch;
- exact-head scratch no-regression guard passed;
- debug cost measured at 1.7426x normal forward on the tiny benchmark.

#90 now only needs the user-facing CLI/debug activation surface tracked in #229.
